### PR TITLE
Adding python API `compute_tensor_descriptor`

### PR DESCRIPTION
**Pull request by jjsjann123**
Added `nvfuser.compute_tensor_descriptor` API.
The new API computes `stride_order` as well as adjusted `contiguity` flag. This is to pave the path to support allocation domain on input in our python integration.

The function sorts input dimensions based on strides before computing contiguity flag.

TODOs in future PR:
1. swap computeContiguity with computeTensorDescriptor in python API `define_tensor`.
2. add stride_order support via allocation domain in `TensorRecord`;
3. remove `compute_contiguity`.

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build

File: csrc/python_frontend/python_bindings.cpp
Comment (zasdfgbnm) [inline csrc/python_frontend/python_bindings.cpp:216]: Can the loop below be merged into this loop by checking `dim_info_vec[last].stride == expect` instead of `== 1`?

File: python_tests/test_python_frontend.py
Comment (zasdfgbnm) [inline python_tests/test_python_frontend.py:1399]: Shouldn't the innermost dim be the expanded dim, which has contiguity `None`?

Comment (jjsjann123) [reply python_tests/test_python_frontend.py:1399]: ha, this is where I wasn't quite sure about how we are representing the contiguity flag.

i.e. are contiguity flag corresponding to logical domain (rfactor domain), or should it be mapping to physical domain (allocation domain)?
Currently I'm viewing it as logical domain, does that make sense to you?

File: csrc/python_frontend/python_bindings.cpp
Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp:216]: will do~

File: python_tests/test_python_frontend.py
Comment (zasdfgbnm) [reply python_tests/test_python_frontend.py:1399]: I think it should be allocation domain?

Comment (jjsjann123) [reply python_tests/test_python_frontend.py:1399]: Thanks for confirming this one. I'll fix it~

File: python_tests/test_python_frontend.py
Comment (jacobhinkle) [inline python_tests/test_python_frontend.py:1390]: I'm trying to understand this example. Is stride order uniquely determined by the sizes and strides when expanded dims are present? In this case wouldn't [5, 4, 2, 1, 3, 0] be a valid stride order also? i.e. any sort of the strides. Do we guarantee a particular order in the cases?

Comment (jjsjann123) [reply python_tests/test_python_frontend.py:1390]: Oops. sorry for the lack of comment. I was trying to confirm I'm returning the right order before writing down any comment/doc.

Comment (jjsjann123) [reply python_tests/test_python_frontend.py:1390]: In short, there's not a unique ordering when we have equal strides.

But we do preserve the original order and rely on that to break tie.

File: csrc/python_frontend/python_bindings.cpp
Comment (jjsjann123) [inline csrc/python_frontend/python_bindings.cpp:187]: @jacobhinkle Does this answer your question?

**Comment by jjsjann123** [comment]
!build

**Review by jacobhinkle — COMMENTED** [review]
Thanks @jjsjann123 the examples help.

File: csrc/python_frontend/python_bindings.cpp
Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp:208]: It's hard for me to argue for what I put in there since I also found it confusing for myself. But I slowly get used to it.

I don't have strong opinions on how we should handle this. But I would like to keep it consistent with potential integration. i.e. whatever framework decided to represent it. :wink:

**Comment by jacobhinkle** [comment]
Please bump version.txt.

File: csrc/python_frontend/python_bindings.cpp
Comment (jacobhinkle) [inline csrc/python_frontend/python_bindings.cpp:216]: `std::pair`

File: csrc/python_frontend/python_bindings.cpp
Comment (jacobhinkle) [inline csrc/python_frontend/python_bindings.cpp:218]: Computing stride order only requires the strides, and contiguity requires all three things: sizes, strides, and stride order. Should we split this into two functions with more descriptive names `computeStrideOrder` and `computeContiguity`?

File: csrc/python_frontend/python_bindings.cpp
Comment (jacobhinkle) [inline csrc/python_frontend/python_bindings.cpp:260]: This would probably be clearer in two loops so that setting stride order doesn't need to be repeated.

File: csrc/python_frontend/python_bindings.cpp
Comment (jacobhinkle) [inline csrc/python_frontend/python_bindings.cpp:136]: Minor point: `isBroadcast` would better match the corresponding `IterDomain` method.

**Review by jacobhinkle — APPROVED** [review]
LGTM. Remember to bump `version.txt`.

File: csrc/python_frontend/python_bindings.cpp
Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp:216]: Any specific reason that we want pair other than a tuple?
I know pair complies to a standard-layout, but I'm don't know if that would add complexity to the pybind handling here.

File: csrc/python_frontend/python_bindings.cpp
Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp:218]: Good catch. We actually still have a `computeContiguity` api that computes contiguity following the logical domain order, rather than trying to compensate the stride order thing. I'm leaving that as-is for now and plan to migrate integration to use the new API.

On the topic of splitting the API. I intend to not do that and keep them as a whole. Hence the decision behind the naming here as `computeTensorDescriptor`.
I'm trying to avoid exposing all the nuts and bolts to user.Hopefully they would treat all the returned value here as an opaque thing; On top of that, `contiguity` vector by itself doesn't make much sense without accompanying the order where it was computed.

File: csrc/python_frontend/python_bindings.cpp
Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp:136]: Fair. let me change that.

File: csrc/python_frontend/python_bindings.cpp
Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp:260]: > so that setting stride order doesn't need to be repeated.

I wasn't quite following the second part. but I will split the for loop. :stuck_out_tongue_winking_eye:

**Comment by jjsjann123** [comment]
Meanwhile, I'm trying to figure out what went wrong in #1021 . If there's codegen issue, exposing that support is going to be... rough...

**Comment by jjsjann123** [comment]
> Meanwhile, I'm trying to figure out what went wrong in #1021 . If there's codegen issue, exposing that support is going to be... rough...

wait... I'm confusing myself with the enablement PR #1031 ... This one is not blocked by anything. I'll merge it after fixing review comments.

File: csrc/python_frontend/python_bindings.cpp
Comment (jacobhinkle) [reply csrc/python_frontend/python_bindings.cpp:216]: > Any specific reason that we want pair other than a tuple? I know pair complies to a standard-layout, but I'm don't know if that would add complexity to the pybind handling here.

Not that I know of other than what you mentioned. I tend to prefer pair to tuple in general just because it's clearer to access the elements and as you say it is guaranteed to be standard layout.

Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp:216]: Let me try that~

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
The failing CI seems to be complaining about driver and it a CI issue?!

I'm merging this as-is.

**Pull request merged**
